### ISSUE TITLE
Strange behavior with union of complex types

### ISSUE THREAD

**Issue by Roaders** [Microsoft/TypeScript#23162]
**TypeScript Version:**  2.8.1

**Search Terms:** 
union complex types 

**Code**

```ts
interface IOne{
    stringOne: string;
    numberOne: number;
}

interface ITwo{
    stringTwo: string;
    numberTwo: number;
}

type IEitherOr = IOne | ITwo;

var itemThree: IEitherOr = {
    numberOne: 42343,
    stringOne: "",

    numberTwo: "sdfsd"
}
```

**Expected behavior:**
This should throw an error.
IMHO I think that there are 3 issues here.

- First off I was surprised that `numberTwo` was allowed AT ALL. I think that the type should be either `IOne` OR `ITwo`, not a mix of the two.
- If we do allowed mixed types (which I don't think we should) then surely if there are any properties for `ITwo` we should require all of them. This should throw an error unless we have specified both `numberTwo` and `stringTwo`
- lastly it's not even type checking `numberTwo`. It is typed as a number but it's fine with a string.

**Actual behavior:**
The compiler is fine with this!

**Playground Link:** 
[Link](http://www.typescriptlang.org/play/#src=%0D%0Ainterface%20IOne%7B%0D%0A%20%20%20%20stringOne%3A%20string%3B%0D%0A%20%20%20%20numberOne%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20ITwo%7B%0D%0A%20%20%20%20stringTwo%3A%20string%3B%0D%0A%20%20%20%20numberTwo%3A%20number%3B%0D%0A%7D%0D%0Atype%20IEitherOr%20%3D%20IOne%20%7C%20ITwo%3B%0D%0A%0D%0Avar%20itemThree%3A%20IEitherOr%20%3D%20%7B%0D%0A%20%20%20%20numberOne%3A%2042343%2C%0D%0A%20%20%20%20stringOne%3A%20%22%22%2C%0D%0A%0D%0A%20%20%20%20numberTwo%3A%20%22sdfsd%22%0D%0A%7D%0D%0A%0D%0Atype%20IFixedEitherOr%20%3D%0D%0A%20%20%20%20(%20IOne%20%26%20Partial%3CITwo%3E%20)%20%0D%0A%20%20%20%20%7C%0D%0A%20%20%20%20(%20ITwo%20%26%20Partial%3CIOne%3E%20)%3B%0D%0A%0D%0Avar%20itemFour%3A%20IFixedEitherOr%20%3D%20%7B%0D%0A%20%20%20%20numberOne%3A%2042343%2C%0D%0A%20%20%20%20stringOne%3A%20%22%22%2C%0D%0A%0D%0A%20%20%20%20numberTwo%3A%20%22dfgdf%22%0D%0A%7D%0D%0A%0D%0Atype%20IBoth%20%3D%20IOne%20%26%20ITwo%3B%0D%0A%0D%0Avar%20itemFive%3A%20IBoth%20%3D%20%7B%0D%0A%20%20%20%20numberOne%3A%2034234%2C%0D%0A%20%20%20%20stringOne%3A%20%22%22%2C%0D%0A%20%20%20%20numberTwo%3A%2034234%2C%0D%0A%20%20%20%20stringTwo%3A%20%22%22%0D%0A%7D)

**Work Around:**

You can enforce proper type checking on the secondary mixed type like this:

```
type IFixedEitherOr =
    ( IOne & Partial<ITwo> ) 
    |
    ( ITwo & Partial<IOne> );

var itemFour: IFixedEitherOr = {
    numberOne: 42343,
    stringOne: "",

    numberTwo: "dfgdf" // this errors
}
```
The code above correctly fails to compile.

**Comment by jack-williams** [comment]
Related: #20863 

There is nothing unsound about this interaction but it can be undesirable which is why there is excess property checking at initialisation, however I believe this does not work for unions without a discriminant (see related issue).

**Comment by jack-williams** [comment]
The type `IEitherOr` does not promise that property `numberTwo` will be a type number. It promises that either:
- `stringOne` will be `string` and `numberOne` will be `number`, OR, 
- `stringTwo` will be `string` and `numberTwo` will be `number`.

If you state `IEitherOr` using intersection and union (which is roughly equivalent), and consider them like logical and, and logic or. Then you have:
- `({stringOne: string} & {numberOne: number}) | ({stringTwo: string} & {numberTwo: number})`

which can be satisfied with only `{stringOne: string}` and `{numberOne: number}`: you never need to claim anything about `numberTwo`.

Note: you **can** get unsound behaviour from this object:
```ts
if ("numberTwo" in itemThree) {
    itemThree.numberTwo // narrows to ITwo, unsafe!
}
```
however this is from narrowing rules for union, not assignment compatibility.
